### PR TITLE
#1788 - Institution student profile

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Institution view student profile

 - [x] Convert the form (studentProfileSummary) to vuetify component
 
image

 - [x] Form definition studentProfileSummary has been removed.
 - [x] Minor refactors done in `student.controller.service.ts` and `StudentRestrictionChip.vue` to get the props typed.
 - [x] The authorization for API will be handled in #1830 as discussed with team. Todo comments added for this authorization part.
 - [x] Routes and Vue files created for other student tabs.
image

### CODE REVIEW

File: sources/packages/backend/libs/test-utils/src/factories/application.ts [hunk 1]
```diff
@@ -49,6 +50,7 @@ export function createFakeApplication(relations?: {
   application.currentAssessment = relations?.currentStudentAssessment;
   application.applicationNumber = faker.random.alphaNumeric(10);
   application.applicationException = relations?.applicationException;
+  application.location = relations?.location;
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application.ts:53]: if the location is not passed, shouldn't we create a fake location as the application status is `applicationStatus = ApplicationStatus.Submitted;` by default

File: sources/packages/backend/libs/test-utils/src/factories/application.ts [hunk 2]
```diff
@@ -154,3 +156,25 @@ export async function saveFakeApplicationDisbursements(
...
+  relations?: {
+    student?: Student;
+    programYear?: ProgramYear;
+    currentStudentAssessment?: StudentAssessment;
+    applicationException?: ApplicationException;
+    institutionLocation?: InstitutionLocation;
+  },
+): Promise<Application> {
+  const applicationRepo = dataSource.getRepository(Application);
+
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application.ts:177]: pls remove the extra line

File: sources/packages/backend/libs/test-utils/src/factories/application.ts [hunk 3]
```diff
@@ -154,3 +156,25 @@ export async function saveFakeApplicationDisbursements(
   savedApplication.currentAssessment = savedOriginalAssessment;
   return applicationRepo.save(savedApplication);
 }
+
+/**
+ * Create and save a fake application.
+ * @param dataSource data source to persist application.
+ * @param relations application entity relations.
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application.ts:163]: it will be better if you could expand the relations

File: sources/packages/backend/libs/test-utils/src/factories/application.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/test-utils/src/factories/application.ts:53]: Not necessarily. As of now we can have application in submitted status and not have a location for it. We have to update the factory when we make location_id as not null.

File: sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.institutions.controller.getStudentProfile.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  // for the institution.
+  it("Should get not found error when student is not found.", async () => {
+    // Arrange
+    const endpoint = "/institutions/student/999999";
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(
+        await getInstitutionToken(InstitutionTokenTypes.CollegeCUser),
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student/_tests_/e2e/student.institutions.controller.getStudentProfile.e2e-spec.ts:99]: https://github.com/bcgov/SIMS/pull/1863/files#r1163061705

File: sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts [hunk 5]
```diff
@@ -45,4 +53,18 @@ export class StudentInstitutionsController extends BaseController {
...
   }
+
+  /**
+   * Get the student information that represents the profile.
+   * TODO: Authorization must be enabled to validate if the student has submitted
+   * at least one application for the institution of the user.
+   * @param studentId student.
+   * @returns student profile details.
+   */
+  @Get(":studentId")
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts:64]: `@ApiNotFoundResponse({`?

File: sources/packages/web/src/components/common/students/StudentProfile.vue [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import { AddressAPIOutDTO } from "@/services/http/dto";
+
+export default defineComponent({
+  props: {
+    studentId: {
+      type: Number,
+      required: true,
+    },
+  },
+  setup(props: any) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/students/StudentProfile.vue:111]: Since this is a new component, can we remove the `any` from the props?

File: sources/packages/backend/libs/test-utils/src/factories/sin-validation.ts [hunk 7]
```diff
@@ -1,8 +1,8 @@
 import { SINValidation, Student } from "@sims/sims-db";
 import * as faker from "faker";
 
-export function createFakeSINValidation(relations: {
-  student: Student;
+export function createFakeSINValidation(relations?: {
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/sin-validation.ts:4]: why optional?

File: sources/packages/web/src/components/common/students/StudentProfile.vue [hunk 8]
```diff
@@ -0,0 +1,129 @@
+<template>
+  <body-header-container>
+    <template #header>
+      <body-header title="Profile" />
+    </template>
+    <p class="category-header-medium">Student profile</p>
```
Comment (ann-aot) [inline sources/packages/web/src/components/common/students/StudentProfile.vue:6]: it's a `H3` according to figma
image

File: sources/packages/web/src/components/common/students/StudentProfile.vue (hunk 6 above)
Comment (ann-aot) [inline sources/packages/web/src/components/common/students/StudentProfile.vue:111]: do we need `any`

File: sources/packages/web/src/services/http/dto/Student.dto.ts [hunk 9]
```diff
@@ -34,10 +34,7 @@ export interface StudentProfileAPIOutDTO {
   pdStatus: StudentPDStatus;
   sin: string;
   sinConsent: boolean;
-}
-
-export interface AESTStudentProfileAPIOutDTO extends StudentProfileAPIOutDTO {
-  hasRestriction: boolean;
+  hasRestriction?: boolean;
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/dto/Student.dto.ts:37]: why is this change? `hasRestriction` is only needed by the ministry right?

File: sources/packages/web/src/types/AppRoutes.ts [hunk 10]
```diff
@@ -47,6 +47,8 @@ export enum AppRoutes {
   InstitutionManageUsers = "manage-users",
   ActiveApplicationScholasticStandingView = "active-application/location/:locationId/scholastic-standing/:scholasticStandingId",
   OfferingsUpload = "offerings-upload",
+  InstitutionStudentDetail = "student-detail/:studentId",
```
Comment (ann-aot) [inline sources/packages/web/src/types/AppRoutes.ts:50]: `student/:studentId` should be enough IMO

File: sources/packages/web/src/types/AppRoutes.ts [hunk 11]
```diff
@@ -47,6 +47,8 @@ export enum AppRoutes {
...
+  InstitutionStudentProfile = "student-profile",
```
Comment (ann-aot) [inline sources/packages/web/src/types/AppRoutes.ts:51]: IMO `profile`

File: sources/packages/web/src/types/contracts/StudentContract.ts [hunk 12]
```diff
@@ -5,14 +5,10 @@ import {
 } from "@/services/http/dto";
 import { IdentityProviders } from "@/types";
 
-export interface StudentFormInfo extends StudentProfileAPIOutDTO {
+export interface StudentProfile extends StudentProfileAPIOutDTO {
   birthDateFormatted: string;
 }
 
-export interface AESTStudentForm extends StudentFormInfo {
```
Comment (ann-aot) [inline sources/packages/web/src/types/contracts/StudentContract.ts:12]: https://github.com/bcgov/SIMS/pull/1863/files#r1163175201

File: sources/packages/web/src/views/aest/student/StudentDetails.vue [hunk 13]
```diff
@@ -13,21 +13,15 @@
                 ? StudentRestrictionStatus.Restriction
                 : StudentRestrictionStatus.NoRestriction
             "
-            :label="
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/StudentDetails.vue:16]: why is this removed?

File: sources/packages/web/src/views/institution/student/InstitutionStudentApplications.vue [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<template>
```
Comment (guru-aot) [inline sources/packages/web/src/views/institution/student/InstitutionStudentApplications.vue:1]: ❤️

**Review by guru-aot — APPROVED** [review]
Thanks for creating the placeholder pages, Please work on the other devs comments. Nice work @dheepak-aot

File: sources/packages/web/src/views/aest/student/StudentDetails.vue [hunk 15]
```diff
@@ -120,21 +111,13 @@ export default defineComponent({
       },
     ]);
 
-    const goBack = () => {
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/StudentDetails.vue:123]: 👍

File: sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue [hunk 16]
```diff
@@ -0,0 +1,100 @@
+<template>
+  <full-page-container layout-template="centered-tab" :full-width="true">
+    <template #header>
+      <header-navigator
+        title="Student Details"
+        :subTitle="studentDetails.fullName"
+      />
+    </template>
+    <template #tab-header>
+      <v-tabs :model="tab" stacked color="primary"
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue:10]: do we need :model="tab"`?

File: sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    onMounted(async () => {
+      studentDetails.value = await StudentService.shared.getStudentProfile(
+        props.studentId,
+      );
+    });
+
+    return {
+      items,
+      studentDetails,
+      StudentRestrictionStatus,
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue:95]: why `StudentRestrictionStatus`?

File: sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+      tab,
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue:96]: I guess, its not required

**Review by ann-aot — COMMENTED** [review]
Good work @dheepak-aot . added some comments

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work and thanks for the placeholders also.
Please take a look at the comments.

File: sources/packages/backend/libs/test-utils/src/factories/student.ts [hunk 19]
```diff
@@ -21,3 +24,21 @@ export function createFakeStudent(user?: User): Student {
...
+
+/**
+ * Create and save fake student
+ * @param dataSource data source to persist student.
+ * @param relations student entity relations.
+ * @returns persisted student with relations provided.
+ */
+export async function saveFakeStudent(
+  dataSource: DataSource,
+  relations?: { user?: User; sinValidation?: SINValidation },
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/backend/libs/test-utils/src/factories/student.ts:36]: Adding SINValidation to student is required for the PR as student profile includes the SIN number in it.

File: sources/packages/backend/libs/test-utils/src/factories/sin-validation.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/test-utils/src/factories/sin-validation.ts:4]: Without making the student optional for SINValidation, we cannot pass a SINValidation object to `saveFakeStudent(relations?: { user?: User; sinValidation?: SINValidation })` because the student of the SINValidation is going to be created inside the saveFakeStudent and SINValidation can be assigned and saved only after the student is persisted, otherwise typeorm is going to throw the cyclic dependency error.

File: sources/packages/web/src/components/common/students/StudentProfile.vue (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/students/StudentProfile.vue:111]: that was missed.

File: sources/packages/web/src/views/aest/student/StudentDetails.vue (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/aest/student/StudentDetails.vue:16]: Apparently there was never a property in `student-restriction-chip` called label and no usage found inside the component.

image

File: sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue (hunk 16 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/institution/student/InstitutionStudentDetails.vue:10]: It has to be removed.

File: sources/packages/web/src/types/AppRoutes.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/AppRoutes.ts:51]: https://github.com/bcgov/SIMS/pull/1863#discussion_r1163338494

**Comment by ann-aot** [comment]
> > > ## Institution view student profile
> > > 
> > > * [x]    Convert the form (studentProfileSummary) to vuetify component
> > > 
> > > image
> > 
> > 
> > I am seeing the red-marked items missing, is it, not part of the PR, image and orange marked wording seems different.
> 
> Yes as discussed these changes are not considered in scope of the story.

just to double check what about the `yellow marked` title? are we not considering in the PR

File: sources/packages/web/src/types/AppRoutes.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/AppRoutes.ts]: Done for both institution and ministry.

File: sources/packages/web/src/types/AppRoutes.ts (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/AppRoutes.ts]: Done for both institution and ministry.

File: sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts:64]: mm, I am not seeing this change

File: sources/packages/web/src/services/http/dto/Student.dto.ts (hunk 9 above)
Comment (ann-aot) [reply sources/packages/web/src/services/http/dto/Student.dto.ts:37]: we can use 2 DTO in the client methods, I believe we did it in many other shared API client methods right? Let me know if I am missing anything

File: sources/packages/backend/libs/test-utils/src/factories/application.ts [hunk 20]
```diff
@@ -154,3 +156,29 @@ export async function saveFakeApplicationDisbursements(
...
+ * - `student` related student.
+ * - `programYear` related program year.
+ * - `currentStudentAssessment` related current assessment.
+ * - `applicationException` related application exception.
+ * - `institutionLocation` related institution location.
+ * @returns persisted application with relations provided.
+ */
+export async function saveFakeApplication(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application.ts:171]: I miss this in the first round of the review, what is the motivation to create this method?
My assumption is that the `saveFakeSimething` would be used when we have a combination of items that must be saved altogether and there is a benefit to grouping those saving in one method. Am I missing something?

File: sources/packages/backend/libs/test-utils/src/factories/student.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/backend/libs/test-utils/src/factories/student.ts:38]: While assigning SINValidation to the student created inside the factory `saveFakeStudent()`, I also wanted to make sure to expose a way of modifying the default sinValidation assigned in the factory from outside even though there is no consumer to it. 

@andrewsignori-aot  @guru-aot  @andrepestana-aot @sh16011993 what is your opinion?

If the team wants to remove I can remove the relations option sinValidation.

**Comment by dheepak-aot** [comment]
> > > > ## Institution view student profile
> > > > 
> > > > * [x]     Convert the form (studentProfileSummary) to vuetify component
> > > > 
> > > > image
> > > 
> > > 
> > > I am seeing the red-marked items missing, is it, not part of the PR, image and orange marked wording seems different.
> > 
> > 
> > Yes as discussed these changes are not considered in scope of the story.
> 
> just to double check what about the `yellow marked` title? are we not considering in the PR

The intention of this ticket is to replicate the details what we have in ministry and any further changes in UI was decided to handled in separate ticket  which involves anything that needs to be changed in the UI for both ministry and institution.

File: sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student/student.institutions.controller.ts:64]: I can see the change. 
image

File: sources/packages/web/src/services/http/dto/Student.dto.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/http/dto/Student.dto.ts:37]: Currently most of the DTOs are not client specific and I was trying to follow that pattern here.

File: sources/packages/backend/libs/test-utils/src/factories/application.ts (hunk 20 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/test-utils/src/factories/application.ts:171]: I got it. Can use save at test case level.

File: sources/packages/backend/libs/test-utils/src/factories/student.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/test-utils/src/factories/student.ts:38]: I do not see a problem to keep as it is.
I am under the impression that we have been following the pattern `relations?.sinValidation ?? createFakeSINValidation(...)` many times.

About "Is our plan to add all possible relations when creating a save", I do not see the reason to make this statement based on this PR. If I am missing something please let me know @ann-aot.

In general, IMO, the factories are helpers targeting the DB data setup. Usually, I would assume that they will be leaning more toward representing the DB than the business. Personally, when I create a new factory I get all the properties from the entity model and I just try to give a nice initial set of values.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, look good 👍

File: sources/packages/backend/libs/test-utils/src/factories/student.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (andrepestana-aot) [reply sources/packages/backend/libs/test-utils/src/factories/student.ts:38]: I don't see any problem leaving that.

[the remark this answers is not in the document]
Comment (ann-aot) [reply sources/packages/backend/libs/test-utils/src/factories/student.ts:38]: As mentioned it was my opinion to use the relations or options or parms on demand, I believe I got some comments supporting that in my old stories. nevermind, as devs are supporting this way. resolving my comment

File: sources/packages/web/src/services/http/dto/Student.dto.ts (hunk 9 above)
Comment (ann-aot) [reply sources/packages/web/src/services/http/dto/Student.dto.ts:37]: As discussed,  we can discuss  it in dev syncup

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot 👍 Thanks for doing the changes 👍
